### PR TITLE
Packaging - systemd-aware RPM spec file

### PR DESCRIPTION
**Pull request by mgrzybek**
According to #128
This PR adds a systemd service file, the required configuration files and renames the "go-carbon.centos.spec" to "go-carbon.spec"

### CODE REVIEW

**Comment by lomik** [comment]
Thanks! I will merge this after 0.9 release

**Comment by xneo64** [comment]
@lomik why is this pushed to 0.10? 
With the USR2 restart feature present in 0.9, I think this PR is essential for the 0.9 release

**Comment by blysik** [comment]
I'm looking to deploy go-carbon onto RHEL7 servers.  Any idea when this will be merged?  Thanks!

**Comment by AlexAkulov** [comment]
- Do you really need `_` prefix in `_graphite` username?
- You should include `network-online.target` in `After` option. Otherwise go-carbon could run before network.
- My config file also include this settings, they seem to be useful
```
TimeoutStopSec=600 # Default value in CentOS is 90s. Because dump saving may take longer, you should increase this number.
LimitNOFILE=55555 # Default limit is 1024. If you don't use relay in front of go-carbon, you might hit this limit.
LimitMEMLOCK=infinity # Cache size may be huge
```

**Comment by mgrzybek** [comment]
The Debian graphite package creates a user called _graphite. That's why I used it there.
I have added your settings.

Moreover I think the configuration directory should be `/etc/carbon` instead of `/etc/go-carbon` to be compliant with graphite-project. Any idea ?

**Comment by lomik** [comment]
> The Debian graphite package creates a user called _graphite. That's why I used it there.
I have added your settings.

Creating a user when installing the package is one of the issues on which I have not formed a final opinion.

> Moreover I think the configuration directory should be /etc/carbon instead of /etc/go-carbon to be compliant with graphite-project. Any idea ?

Projects are independent and have incompatible configs. Moreover, you can run them simultaneously.

**Comment by hnakamur** [comment]
I'm using go-carbon on CentOS 7. This RPM package would be great!
Thanks for this pull request!

> Creating a user when installing the package is one of the issues on which I have not formed a final opinion.

I think it's fairly common for packages to create a user when installed
so many people would expect a user is created.
However I disagree `_` prefix in user names for rpm packages since it's uncommon.
In CentOS 7, use names without `_` prefix such as `nginx` or `apache` are used.

**Comment by hnakamur** [comment]
Thanks for changing username from _graphite to carbon for RPM!

File: deploy/go-carbon.spec [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+Group:		Development/Tools
+License:	MIT License
+URL:		https://github.com/lomik/go-carbon
+Source0:	%{name}.tar.gz
+BuildRoot:      %{_tmppath}/%{name}-%{version}
+
+%if systemd_requires
+%systemd_requires
+%else
+Requires:  /etc/init.d/functions
```
Comment (ehlerst) [inline deploy/go-carbon.spec]: Can we change this to "initscripts" this breaks yum dependencies  in amazon linux.

```
Error: Package: go-carbon-0.9.1-1.el6.x86_64 (@timsrepo-rpm-repo)
           Requires: /etc/init.d/functions
           Removing: initscripts-9.03.49-1.34.amzn1.x86_64 (installed)
               Not found
           Updated By: initscripts-9.03.49-1.35.amzn1.x86_64 (amzn-updates)
               Not found
yum whatprovides /etc/init.d/functions
Loaded plugins: priorities, s3iam, update-motd, upgrade-helper
initscripts-9.03.49-1.34.amzn1.x86_64 : The inittab file and the /etc/init.d scripts
Repo        : installed
Matched from:
Filename    : /etc/init.d/functions
```

File: deploy/go-carbon.spec [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+%{__install} -pD -m 755 %{name} %{buildroot}/%{_bindir}/%{name}
+%{__install} -pD -m 644 %{name}.conf %{buildroot}/%{_sysconfdir}/%{name}/%{name}.conf
+%{__install} -pD -m 644 storage-schemas.conf %{buildroot}/%{_sysconfdir}/%{name}/storage-schemas.conf
+%{__install} -pD -m 644 storage-aggregation.conf %{buildroot}/%{_sysconfdir}/%{name}/storage-aggregation.conf
+
+%if systemd_requires
+%{__install} -pD -m 644 %{name}.service %{buildroot}%{_unitdir}/%{name}.service
+%else
+%{__install} -pD -m 755 %{name}.init %{buildroot}/etc/init.d/%{name}
```
Comment (ehlerst) [inline deploy/go-carbon.spec]: use the %{_initddir} macro here instead of /etc/init.d

**Comment by lomik** [comment]
I think to keep demonization and user configuration in go-carbon. Otherwise systemd should create and chown log, data and other directories (and user should support same settings in go-carbon config and systemd service file).

I'm testing this systemd service file
```
[Unit]
Description=Golang implementation of Graphite/Carbon server.
Documentation=https://github.com/lomik/go-carbon
After=network.target

[Service]
Type=forking
ExecStart=/usr/bin/go-carbon -config /etc/go-carbon/go-carbon.conf -pidfile /var/run/go-carbon.pid -daemon
ExecReload=/bin/kill -HUP $MAINPID
KillSignal=USR2
Restart=on-failure

TimeoutStopSec=600
LimitNOFILE=55555
LimitMEMLOCK=infinity

[Install]
WantedBy=multi-user.target
```

**Comment by ehlerst** [comment]
@lomik if you do -daemon the pid returns an exit 0 and systemd gets confused killing the pid and trying to restart it. I had tried this before submitting my merge for the same, i think they expect everything in the foreground. 

Edit: its from my systemd merge but its still the same problem.
ExecStart=/usr/local/bin/go-carbon -config /usr/local/etc/go-carbon.conf -pidfile /var/run/go-carbon.pid -daemon

 ```
go-carbon.service - go-carbon a golang implementation of Graphite/Carbon server
   Loaded: loaded (/usr/lib/systemd/system/go-carbon.service; disabled; vendor preset: disabled)
   Active: failed (Result: start-limit) since Tue 2017-04-04 17:26:44 UTC; 2s ago
  Process: 22041 ExecStart=/usr/local/bin/go-carbon -config /usr/local/etc/go-carbon.conf -pidfile /var/run/go-carbon.pid -daemon (code=exited, status=0/SUCCESS)
 Main PID: 22041 (code=exited, status=0/SUCCESS)

Apr 04 17:26:44 : go-carbon.service holdoff time over, scheduling restart.
Apr 04 17:26:44 : start request repeated too quickly for go-carbon.service
Apr 04 17:26:44 : Failed to start go-carbon a golang implementation of Graphite/Carbon server.
Apr 04 17:26:44 : Unit go-carbon.service entered failed state.
Apr 04 17:26:44 : go-carbon.service failed.
```

Without -daemon

```
 systemctl start go-carbon
 systemctl status go-carbon
● go-carbon.service - go-carbon a golang implementation of Graphite/Carbon server
   Loaded: loaded (/usr/lib/systemd/system/go-carbon.service; disabled; vendor preset: disabled)
   Active: active (running) since Tue 2017-04-04 17:29:16 UTC; 5s ago
 Main PID: 22116 (go-carbon)
   CGroup: /system.slice/go-carbon.service
           └─22116 /usr/local/bin/go-carbon -config /usr/local/etc/go-carbon.conf

Apr 04 17:29:16  systemd[1]: Started go-carbon a golang implementation of Graphite/Carbon server.
Apr 04 17:29:16  systemd[1]: Starting go-carbon a golang implementation of Graphite/Carbon server...
Apr 04 17:29:18 systemd[1]: Started go-carbon a golang implementation of Graphite/Carbon server.
```

**Comment by lomik** [comment]
@ehlerst,
Do you set Type=forking?

**Comment by tehlers320** [comment]
@lomik nope! i retract my comments. it works with Type=forking.

File: deploy/go-carbon.spec (hunk 1 above)
Comment (mgrzybek) [reply deploy/go-carbon.spec]: Change made.

**Pull request closed without merging**